torch/testing/_internal/common_utils.py

```diff
@@ -69,6 +69,116 @@
+        ops, all modules, or all ops + their associated dtypes.
+
+        Args:
+            test (fn): Test function to parametrize over
+            generic_cls (class): Generic test class object containing tests (e.g. TestFoo)
+            device_cls (class): Device-specialized test class object (e.g. TestFooCPU); set to None
+                if the tests are not part of a device-specific set
+
+        Returns:
+            Generator object returning 3-tuples of:
+                test (fn): Parametrized test function; must support a device arg and args for any params
+                test_name (str): Parametrized name of the test (e.g. test_bar_opname_int64)
+                param_kwargs (dict): Param kwargs to pass to the test (e.g. {'op': 'add', 'dtype': torch.int64})
+        """
+        raise NotImplementedError
+
+    def __call__(self, fn):
+        fn.parametrize_fn = self._parametrize_test
+        return fn
+
+
+def instantiate_parametrized_tests(generic_cls):
+    """
+    Instantiates tests that have been decorated with a parametrize_fn. This is generally performed by a
+    decorator subclass of _TestParametrizer. The generic test will be replaced on the test class by
+    parametrized tests with specialized names.
+
+    Args:
+        generic_cls (class): Generic test class object containing tests (e.g. TestFoo)
+    """
+    for attr_name in tuple(dir(generic_cls)):
+        class_attr = getattr(generic_cls, attr_name)
+        if not hasattr(class_attr, 'parametrize_fn'):
+            continue
+
+        # Remove the generic test from the test class.
+        delattr(generic_cls, attr_name)
+
+        # Add parametrized tests to the test class.
+        def instantiate_test_helper(cls, name, test, param_kwargs):
+            @wraps(test)
+            def instantiated_test(self, param_kwargs=param_kwargs):
+                test(self, **param_kwargs)
+
+            assert not hasattr(generic_cls, name), "Redefinition of test {0}".format(name)
+            setattr(generic_cls, name, instantiated_test)
+
+        for (test, test_name, param_kwargs) in class_attr.parametrize_fn(
+                class_attr, generic_cls=generic_cls, device_cls=None):
+            instantiate_test_helper(cls=generic_cls, name=test_name, test=test, param_kwargs=param_kwargs)
+
+
+class parametrize(_TestParametrizer):
+    """
+    Decorator for applying generic test parametrizations.
+
+    Args:
+        arg_str (str): String of arg names separate by commas (e.g. "x,y")
+        arg_values (iterable): Iterable of arg values (e.g. range(10)) or
+            tuples of arg values (e.g. [(1, 2), (3, 4)])
+    """
+    def __init__(self, arg_str, arg_values):
+        self.arg_names = arg_str.split(',')
+        self.arg_values = arg_values
+
+    def _parametrize_test(self, test, generic_cls, device_cls):
+
+        # Build a single composite test that tests all the parametrized cases.
```

Comment:
How important is each variant being instantiated in its own test, @zou3519?

Reply:
It's pretty important but I wouldn't block this PR on this. FWIW pytest also generates a test case per variant.

The main argument for each variant getting its own test is because it makes it easier to tell exactly what variant a test case failed for. For example, if I have some autograd tests that are parametrized over being for "reverse-mode" or "forward-mode" AD (this happens in some tests in test/test_autograd.py) if a test fails I want to be able to know if reverse-mode is broken, if forward-mode is broken, or if both are broken.

Reply:
Definitely need variants as individual tests, I agree! See my comment above for how pytest does this, from what I can tell. Given that, there's a UX discussion to be had on the right way to allow devs to name their variants.